runtime/src/main/scala/akka/grpc/internal/ClientState.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.grpc.internal
+
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.grpc.GrpcClientSettings
+import akka.stream.{ ActorMaterializer, Materializer }
+import io.grpc.ManagedChannel
+
+import scala.annotation.tailrec
+import scala.concurrent.{ ExecutionContext, Future, Promise }
+import scala.util.Failure
+
+/**
+ * INTERNAL API
+ * Used from generated code so can't be private.
+ *
+ * Client utilities taking care of Channel reconnection and Channel lifecycle in general.
+ */
+@InternalApi
+class ClientState(settings: GrpcClientSettings)(implicit mat: Materializer, ex: ExecutionContext) {
+
+  private val internalChannelRef: AtomicReference[InternalChannel] = new AtomicReference[InternalChannel](create())
+  private val closeDemand: Promise[Done] = Promise[Done]()
+
+  mat match {
+    case m: ActorMaterializer =>
+      m.system.whenTerminated.foreach(_ => close())(ex)
```

Comment:
But what about when the materializer is shut down, or can it not be because we create it internally elsewhere?

Also, if the system was terminated, will that `ex` really be something you can schedule jobs on?

Reply:
Good questions.  But, for the record, this is just moved code: https://github.com/akka/akka-grpc/blob/v0.4.1/codegen/src/main/twirl/templates/ScalaClient/Client.scala.txt#L34-L38

Reply:
So should we remove this then?